calyx/src/analysis/dominator_map.rs

```diff
@@ -0,0 +1,365 @@
+            }
+            ir::Control::If(_) => {
+                hs.insert(Self::get_guaranteed_attribute(c, END_ID));
+            }
+            ir::Control::Seq(ir::Seq { stmts, .. }) => {
+                match (&stmts[..]).last() {
+                    None => (),
+                    Some(control) => return Self::get_final(control),
+                }
+            }
+            ir::Control::Par(ir::Par { stmts, .. }) => {
+                for stmt in stmts {
+                    let stmt_final = Self::get_final(stmt);
+                    hs = hs.union(&stmt_final).copied().collect()
+                }
+            }
+        }
+        hs
+    }
+
+    //Given an id and its predecessors pred, and a domination map d_map, updates
+    //d_map accordingly (i.e. the union of all dominators of the predecessors),
+    //plus itself.
+    fn update_node(pred: &HashSet<u64>, id: u64, d_map: &mut DominatorMap) {
+        let mut pred_dominators: Vec<&HashSet<u64>> = Vec::new();
+        for id in pred.iter() {
+            if let Some(dominators) = d_map.map.get(id) {
+                pred_dominators.push(dominators);
+            }
+        }
+        let mut union: HashSet<u64> = HashSet::new();
+        pred_dominators.iter().for_each(|set| {
+            union = union.union(set).copied().collect();
+        });
+        union.insert(id);
+        d_map.map.insert(id, union);
+    }
+
+    //Looks through each "node" in the "graph" and updates the dominators accordingly
+    fn update_map(
+        main_c: &ir::Control,
+        cur_id: u64,
+        pred: &HashSet<u64>,
+        d_map: &mut DominatorMap,
+    ) {
+        let c = match Self::get_control(cur_id, main_c) {
+            Some(control) => control,
+            None => return,
+        };
+        match c {
+            ir::Control::Empty(_) => {
+                unreachable!(
+                    "should not pattern match agaisnt empty in update_map()"
+                )
+            }
+            ir::Control::Invoke(_) | ir::Control::Enable(_) => {
+                Self::update_node(pred, cur_id, d_map);
+            }
+            ir::Control::Seq(ir::Seq { stmts, .. }) => {
+                //Should find a better way of doing this. This is ugly.
+                let mut cur_pred = HashSet::new();
+                let mut first = true;
+                for stmt in stmts {
+                    let id = Self::get_id(stmt);
+                    if first {
+                        Self::update_map(main_c, id, pred, d_map);
+                    } else {
+                        Self::update_map(main_c, id, &cur_pred, d_map);
+                    }
+                    //If the current stmt is just an emtpy par block, for example,
+                    // then we don't want to update the predecessors
+                    let possible_final = Self::get_final(stmt);
+                    if !possible_final.is_empty() {
```

Comment:
Make this code simpler by assuming that you never encounter an empty par or seq block